server/core_storage.go

```diff
@@ -646,20 +646,16 @@ func storagePrepBatch(batch *pgx.Batch, authoritativeWrite bool, op *StorageOpWr
 		// allows us to fetch row state after update even if update itself fails WHERE
 		// condition.
 		// That is returned values are final state of the row regardless of UPDATE success
-		// Order by is necessary as UNION ALL may not guarantee ordering of results, but it is required
-		// to ensure that if update happens values from said update are returned.
-		// At most 2 rows are ever returned so ordering operation should be negligible.
 		query = `
 		WITH upd AS (
 			UPDATE storage SET value = $4, version = $5, read = $6, write = $7, update_time = now()
 			WHERE collection = $1 AND key = $2 AND user_id = $3 AND version = $8
 		` + writeCheck + `
 			RETURNING read, write, version, create_time, update_time
 		)
-		(SELECT read, write, version, create_time, update_time, true AS update FROM upd WHERE EXISTS (SELECT 1 FROM upd))
+		(SELECT read, write, version, create_time, update_time, true AS update FROM upd)
 		UNION ALL
-		(SELECT read, write, version, create_time, update_time, false AS update FROM storage WHERE collection = $1 and key = $2 and user_id = $3 AND NOT EXISTS (SELECT 1 FROM upd))
-		LIMIT 1`
+		(SELECT read, write, version, create_time, update_time, false AS update FROM storage WHERE collection = $1 and key = $2 and user_id = $3 AND NOT EXISTS (SELECT 1 FROM upd))`
```

Comment:
`LIMIT 1` is still handy as it prevents whole branch of `UNION ALL` from exeucting if update succeeded. Without it it is always executed, but returns no rows on update success. End result is the same, but not executing should be ever so slightly cheaper on CPU .

You should be able to verify it with `EXPLAIN ANALYZE`, where `LIMIT 1` version shows  `(never executed)`   part of the plan.